Fix of 1211 and 1082

bingbing8 opened the pull request:
1. fix of https://github.com/PowerShell/Win32-OpenSSH/issues/1211
2. fix of https://github.com/PowerShell/Win32-OpenSSH/issues/1082
3. scp works when powershell and cmd as default shell
3. have the escaping logic in posix_spawn only
4. Add program dir as part of path in process context so scp and sftp in…
5. only escape double quotes and backslash for common shells

bingbing8 committed:
Fix of 1211 and 1082

1. fix of 1211 (scp works when powershell and cmd as default shell)
2. fix of 1082,
3. have the escaping logic in posix_spawn only4.
Add program dir as part of path in process context so scp and sftp in…

bingbing8 committed:
add quotes for the cmd only when module path is added.

bingbing8 committed:
fix the cmd when it already surrounded by quotes

bingbing8 committed:
don't escape command for cmd.exe and shellhost.exe

bingbing8 committed:
minor updates

bingbing8 committed:
CreateProcess does not support app surrounded by single quotes.

bingbing8 committed:
added tests for double quotes escaping

bingbing8 committed:
update tests

bingbing8 committed:
remove unused var

bingbing8 committed:
minor changes

manojampalam commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1318,7 +1318,7 @@ int
 {
 	int retVal = 0;
-	if(*path == '\"') /* skip double quote if path is "c:\abc" */
+	if(*path == '\"' || *path == '\'') /* skip double quote if path is "c:\abc" */
```
Looks like 'tmp'.exe is valid on Windows. Are we intentionally doing this to benefit the general case ? If so, we may have to document this limitation somewhere.

bingbing8 replied:
I don't get your point here.  how 'tmp'.exe related with absolute path?

manojampalam replied:
never mind. You're right. they are not related.

manojampalam commented on contrib/win32/win32compat/misc.c:
```diff
@@ -1584,198 +1584,70 @@ get_sid(const char* name)
+	* we want to launch scp and sftp executables from the same binary directory
+	* that sshd is hosted in. This will facilitate hosting and evaluating
+	* multiple versions of OpenSSH at the same time.
+	*/
```
looks like this function is not prepending the current binary location to command. are these comments obsolete ?

bingbing8 replied:
1. we added current binary location to PATH
2. If we prepend the path, some shells, like powershell will require shell specific additional characters (& for powershell) to be able to run scp, sftp. powershell -c "& `<full path to scp`>"

bingbing8 replied:
I will remove the comments

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1052,7 +1053,9 @@ spawn_child_internal(char* cmd, char *const argv[], HANDLE in, HANDLE out, HANDL
 		return ret;
 	}
-	
+	if (strstr(path, "system32\\cmd") || strstr(path, "ssh-shellhost"))
```
What if you do:
bash.exe -c c:\windows\system32\cmd.exe /c "echo hi"

bingbing8 replied:
Good catch.

bingbing8 replied:
this case is an issue when the path is user define cmd which contains both exe and arguments. in next version, I use the last parameter to define if it is user defined cmd. If so, we don't change anything.

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1090,54 +1121,108 @@ spawn_child_internal(char* cmd, char *const argv[], HANDLE in, HANDLE out, HANDL
-		memcpy(t, tmp, strlen(tmp));
-		t += strlen(tmp);
-	} else {
+		tmp += strlen(exe_extenstion); /* move the pointer to the end of ".exe" */		
```
can you end up with an input command like the following here:

  c:\windows\system32\cmd /c ping.exe localhost

If so, this logic would incorrectly identify the nested binary as the parent executable

bingbing8 replied:
For the existing known shells and ssh executable, the first parts always ends with .exe. I don't see other ways to identify the file path (the first part)  in the string.

manojampalam replied:
What about with proxyCommand or SSH_ASK_PASS ? Also fix spelling mistake in exe_extenstion.

bingbing8 replied:
Yes, you are right.

bingbing8 replied:
in next version, I use the last parameter to define if it is user defined cmd. If so, we don't change anything. we expect the cmd user enter already have right quotes and escaping

manojampalam commented on contrib/win32/win32compat/wmain_sshd.c:
```diff
@@ -260,14 +260,40 @@ int argc_original = 0;
+			error("failed to allocation memory");
+			return -1;
+		}
+		swprintf_s(path_new_value, path_new_len, L"%s%s%s", __wprogdir, path_value ? L";" : L"",  path_value);
```
What's the confirmed order of exe lookup (we figured that what's on msdn is wrong) ? Does this guarantee picking up exe from progdir in all scenarios?
Are we OK with adjusting PATH for user's remote sessions?

bingbing8 replied:
the msdn is correct, but the search order is for the executable, which is the first part of the command. but scp, sftp are the arguments part of the command.
So far, looks like it picked up binaries from progdir in all the scenarios.
I don't see a problem with adjusting PATH for user's remote session.

bingbing8 committed:
PAL of do_exec_* APIs

bingbing8 committed:
1. change file names
2. don't touch user defined cmd: no prepending, no escaping, no additional quotes
3. fix line ending

bingbing8 committed:
fix file name

bingbing8 commented:
@manojampalam , when you get time, please review the changes. Thanks!

bingbing8 committed:
Merge branch 'latestw_all' of https://github.com/powershell/openssh-portable into cmdescaping

bingbing8 committed:
only escape double quotes and backslash for powershell, bash, and cygwin

bingbing8 committed:
no -c  or /c for shells other than cmd,  ps, bash, and cygwin

bingbing8 committed:
add -c for ssh-shellhost

bingbing8 committed:
minor update

bingbing8 commented:
@manojampalam Please review. currently only escape non-user defined commands from common shells. for other shells, no escaping, no -c added if shell option is not specified

bingbing8 committed:
update comments

bingbing8 committed:
update comments

bingbing8 committed:
test updates

bingbing8 committed:
print out the file for test investigation

bingbing8 committed:
Caller of posix_spawn_internal determine if escaping needed.

bingbing8 committed:
define shell types

manojampalam commented on contrib/win32/openssh/ssh.vcxproj:
```diff
@@ -496,6 +496,7 @@
     <ClInclude Include="$(OpenSSH-Src-Path)uuencode.h" />
     <ClInclude Include="$(OpenSSH-Src-Path)version.h" />
     <ClInclude Include="$(OpenSSH-Src-Path)xmalloc.h" />
+    <ClInclude Include="$(OpenSSH-Src-Path)pal-doexec.c.h" />
```
pal-doexec.h

bingbing8 replied:
removed

manojampalam commented on contrib/win32/openssh/sshd.vcxproj:
```diff
@@ -1,473 +1,474 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Fix CRLF

bingbing8 replied:
fixed

manojampalam replied:
still shows up different

bingbing8 replied:
I changed the whole file to CRLF. Not sure why it is still showing here

bingbing8 replied:
finally fixed it.

manojampalam commented on contrib/win32/win32compat/misc_internal.h:
```diff
@@ -70,9 +70,9 @@ int is_absolute_path(const char *);
 PSID get_sid(const char*);
 int am_system();
-char* build_session_commandline(const char *, const char *, const char *);
+char * build_command_string(const char * );
```
you could get rid of this since this is internal to logic within w32_doexec.c

bingbing8 replied:
removed

bingbing8 replied:
Added back and keep them in misc_internal.h. we need to call them from w32_doexec.c and unit tests

manojampalam commented on contrib/win32/win32compat/misc_internal.h:
```diff
@@ -70,9 +70,9 @@ int is_absolute_path(const char *);
+char * build_command_string(const char * );
 int is_conpty_supported();
-int exec_command_with_pty(wchar_t*, STARTUPINFOW*, PROCESS_INFORMATION*, int);
+int exec_command_with_pty(int * pid, char* cmd, int in, int out, int err, unsigned int col, unsigned int row, int ttyfd);
```
This one too, you could get rid of this since this is internal to logic within w32_doexec.c

bingbing8 replied:
removed

bingbing8 replied:
I revert it back to misc.internal.h because the definition is in w32_pty.c and the caller is w32_doexe.c

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1069,60 +1070,154 @@ spawn_child_internal(char* cmd, char *const argv[], HANDLE in, HANDLE out, HANDL
-		t += strlen(path);
-
+	
+	/*For user defined cmd, no need to add quotes*/
```
?? can you add some examples? What's a user defined command ?

bingbing8 replied:
added

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1069,60 +1070,154 @@ spawn_child_internal(char* cmd, char *const argv[], HANDLE in, HANDLE out, HANDL
+			t += strlen(__progdir);
+			*t++ = '\\';
+		}
+		/*
```
I'm not in favor of this. What issues are we hitting if we don't do this? If the module path has spaces in it, it is expected to be enclosed in quotes. This is a standard practice even on Unix.

bingbing8 replied:
removed

bingbing8 replied:
remove the log below this.

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1069,60 +1070,154 @@ spawn_child_internal(char* cmd, char *const argv[], HANDLE in, HANDLE out, HANDL
-		while (*t1)
-			cmdline_len += (DWORD)strlen(*t1++) + 1 + 2;
+		while (*t1) {
+			if (!escape)
```
Recommend pulling all the commandline building logic into a separate utility helper and writing an exhaustive set of unit tests.

bingbing8 replied:
the is for the case when the module path is added in this function. We need to add quotes

bingbing8 replied:
separate the utility helper out. 
added unit tests.

manojampalam commented on contrib/win32/win32compat/win32_pty.c:
```diff
@@ -96,33 +98,62 @@ int is_conpty_supported()
+
+	memset(&si, 0, sizeof(STARTUPINFO));
+	si.cb = sizeof(STARTUPINFO);
+	si.dwXSize = 5;
```
Why 5 ? Add comments if needed

bingbing8 replied:
remove these two lines. It was from the existing codes.

manojampalam commented on contrib/win32/win32compat/wmain_sshd.c:
```diff
@@ -224,14 +224,40 @@ int argc_original = 0;
 	/* change current directory to sshd.exe root */
 	_wchdir(__wprogdir);
 
+	_wdupenv_s(&path_value, &len, L"PATH");
```
Please add comments on why we are doing this

bingbing8 replied:
added.

manojampalam reviewed (changes requested):
comments inline...

bingbing8 committed:
1. separate cmdline building log from posix_spawn
2. Add registry key to indicate escaping is needed or not for shells other than ps, shellhost, cmd, cygwin, bash. Default is escape the arg string

bingbing8 committed:
add unit tests for build_exec_command and build_commandline_string

bingbing8 committed:
merge conflicts

bingbing8 committed:
revert the comment of one test

bingbing8 committed:
Always add quotes if they are not already there for executables unless createprocess failed. Remove the added quotes as failoever when arg is empty.

bingbing8 committed:
check failover condition in posix_spawn

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1034,115 +1034,25 @@ int fork()
-	*t = '\0';
-
+	int ret = -1;
+	cmdline = build_commandline_string(cmd, argv, prepend_module_path);
```
what if build_commandline_string fails ?

bingbing8 replied:
I will add check

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1034,115 +1034,25 @@ int fork()
 	verbose("fork is not supported"); 
 	return -1;
 }
+char * build_commandline_string(const char* cmd, char *const argv[], BOOLEAN prepend_module_path);
```
can this move to misc_internal.h

bingbing8 replied:
sure

manojampalam commented on contrib/win32/win32compat/w32fd.c:
```diff
@@ -1154,13 +1064,19 @@ spawn_child_internal(char* cmd, char *const argv[], HANDLE in, HANDLE out, HANDL
+		else
+			b = CreateProcessW(NULL, t, NULL, NULL, TRUE, flags, NULL, NULL, &si, &pi);
+		*(cmdline_utf16 + wcslen(cmdline_utf16) - 1) = L'\0';
+		/*failover only when double quotes is potentially added necessary*/
```
you could simplify the below condition by putting this inside loop and getting rid of "num" and "t":

if  ( !b && GetLastError() != ERROR_FILENOTFOUND && (argv == NULL || *argv == NULL) && cmd[0] != '\"') {
        cmdline_utf16++;
        *(cmdline_utf16 + wcslen(cmdline_utf16) - 1) = L'\0';
}

break;

bingbing8 replied:
I can get rid of num, but not t.
I need to free cmdline_utf16 at the end. the pointer can't change

bingbing8 replied:
I can't get rid of num since I need to make sure only one more time is tried

manojampalam commented on contrib/win32/win32compat/win32_pty.c:
```diff
@@ -96,33 +98,60 @@ int is_conpty_supported()
 	debug3("pty commandline: %ls", pty_cmdline);
 
-	if (!CreateProcessW(NULL, pty_cmdline, NULL, NULL, TRUE, 0, NULL, NULL, si, pi)) {
+	if (CreateProcessW(NULL, pty_cmdline, NULL, NULL, TRUE, 0, NULL, NULL, &si, &pi)) {
```
Sorry, I may have missed this earlier but why can't we call posix_spawnp here ?

bingbing8 replied:
for pty, the si setting from session (col, row, and dwflag) are different from posix_spawnp. There is no args for col, row, and dwflag to posix_spawnp. One option is change the signature of posix_spawn_internal to take si info and pass them to spawn_child_internal. and then call posix_spawn_internal at here.

manojampalam commented on regress/unittests/win32compat/miscellaneous_tests.c:
```diff
@@ -317,100 +317,101 @@ test_chroot()
-	out = build_session_commandline("c:\\powershell", NULL, "sftp-server -arg");
-	ASSERT_STRING_EQ(out, "\"c:\\powershell\" -c \"sftp-server.exe -arg\"");
+char *
+build_commandline_string(const char* cmd, char *const argv[], BOOLEAN prepend_module_path);
```
you can get rid of this by moving it to misc_internal

bingbing8 replied:
sure

bingbing8 committed:
Fix the case when the path has quotes and and add_module_path

bingbing8 committed:
change line ending

bingbing8 committed:
changed encoding

bingbing8 committed:
update line ending and encoding

bingbing8 committed:
revert the new line in sshd.vcproj.filters

bingbing8 commented:
@manojampalam, please review the updated code

bingbing8 committed:
minor update for format
